creator-node/src/middlewares.js

```diff
@@ -389,6 +436,118 @@ async function getCreatorNodeEndpoints ({ serviceRegistry, logger, wallet, block
+    const MaxRetries = 201
+    const RetryTimeout = 1000 // 1 seconds
+
+    let blockNumberIndexed = false
+    for (let retry = 1; retry <= MaxRetries; retry++) {
+      logger.info(`${logPrefix} retry #${retry}/${MaxRetries} || time from start: ${Date.now() - start}. Polling until blockNumber ${blockNumber}.`)
+
+      try {
+        // will throw error if blocknumber not found
+        replicaSet = await libs.contracts.UserReplicaSetManagerClient.getUserReplicaSetAtBlockNumber(userId, blockNumber)
+        blockNumberIndexed = true
+        break
+      } catch (e) { } // Ignore all errors until MaxRetries exceeded
+
+      await utils.timeout(RetryTimeout)
+    }
+
+    // Error if failed to retrieve replicaSet
+    if (!replicaSet || !replicaSet.hasOwnProperty('primaryId') || !replicaSet.primaryId) {
+      throw new Error(`${logPrefix} Failed to retrieve user from UserReplicaSetManager after ${MaxRetries} retries. Aborting.`)
+    }
+
+    // Error if failed to index target blockNumber
+    if (!blockNumberIndexed) {
+      throw new Error(`${logPrefix} Web3 provider failed to index target blockNumber ${blockNumber} after ${MaxRetries} retries. Aborting.`)
+    }
+  } else if (ensurePrimary && selfSpID) {
+    /**
+     * If ensurePrimary required but no blockNumber provided, poll URSM until returned primary = selfSpID
+     * Error if still mismatched after specified timeout
+     */
+
+    // Will poll every second for 60 sec
+    const MaxRetries = 61
+    const RetryTimeout = 1000 // 1 sec
+
+    for (let retry = 1; retry <= MaxRetries; retry++) {
+      logger.info(`${logPrefix} retry #${retry}/${MaxRetries} || time from start: ${Date.now() - start}. Polling until primaryEnsured.`)
+
+      try {
+        replicaSet = await libs.contracts.UserReplicaSetManagerClient.getUserReplicaSet(userId)
+
+        if (replicaSet && replicaSet.hasOwnProperty('primaryId') && replicaSet.primaryId === selfSpID) {
+          break
+        }
+      } catch (e) { } // Ignore all errors until MaxRetries exceeded
+
+      await utils.timeout(RetryTimeout)
+    }
+
+    // Error if failed to retrieve replicaSet
+    if (!replicaSet || !replicaSet.hasOwnProperty('primaryId') || !replicaSet.primaryId) {
+      throw new Error(`${logPrefix} Failed to retrieve user from UserReplicaSetManager after ${MaxRetries} retries. Aborting.`)
+    }
+
+    // Error if returned primary spID does not match self spID
+    if (replicaSet.primaryId !== selfSpID) {
+      throw new Error(`${logPrefix} After ${MaxRetries} retries, found different primary (${replicaSet.primaryId}) for user. Aborting.`)
+    }
+  } else {
+    /**
+     * If neither of above conditions are met, falls back to single discprov query without polling
+     */
+
+    logger.info(`${logPrefix} ensurePrimary = false, fetching user replicaSet without retries`)
+
+    try {
+      replicaSet = await libs.contracts.UserReplicaSetManagerClient.getUserReplicaSet(userId)
+    } catch (e) { }
```

Comment:
worth logging here too? it's a different kind of error the way it's written. would be also down to clean this up so the throw is exactly the failure case we care about

Reply:
yeah i can clean this up to distinguish error vs user replica set not present